kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/int-simplification.k

```diff
@@ -151,11 +155,15 @@ module INT-SIMPLIFICATION-COMMON
     rule A /Int 1 => A  [simplification, preserves-definedness]
 
     // safeMath mul check c / a == b where c == a * b
-    rule (A *Int B) /Int A => B requires A =/=Int 0 [simplification, preserves-definedness]
+    rule (A *Int B) /Int A => B requires notBool A ==Int 0 [simplification, preserves-definedness]
+    rule (B *Int A) /Int A => B requires notBool A ==Int 0 [simplification, preserves-definedness]
 
-    rule ((A *Int B) /Int C) /Int B => A /Int C requires B =/=Int 0 andBool C =/=Int 0 [simplification, preserves-definedness]
+    rule ((A *Int B) /Int C) /Int B => A /Int C requires notBool B ==Int 0 andBool notBool C ==Int 0 [simplification, preserves-definedness]
```

Comment:
Hmmmm, is this one true in the `Int` domain? Could we have rounding errors in `(A *Int B) /Int C`, that result in a value that is no longer divisible by `B`?

say:

- `A == 15`
- `B == 4`
- `C == 7`
- `(A * B) / C == 60 / 7 == 8`
- `((A * B) / C) / B == 8 / 4 == 2`
- `A / C == 2`

Hmmm, I can't come up with an example where it's violated. Could we ask an SMT solver if it's true? For lemmas we weren't sure about, we used to provide inline solver transcripts: https://github.com/runtimeverification/evm-semantics/blob/b3198a80e462a13423cf64cf7d5a7f5e347415b4/tests/specs/mcd/verification.k#L69

Reply:
Another caveat with division, should ew require the numbers to be non-negative? Is `/Int` rounding towards zero, or rounding towards nearest?

It looks like `/Int` and `%Int` use `t-division`, which rounds towards zero and `divInt` and `modInt` use Euclidian division, which rounds towards nearest. Does that affect these lemmas? https://github.com/runtimeverification/k/blob/master/k-distribution/include/kframework/builtin/domains.md#integer-arithmetic

Reply:
`/Int`, as far as I know, is the mathematical floor function.

Reply:
If it is not floor, then it's definitely safer to go non-negative.

Reply:
Ok, so, Solidity rounds towards zero, presumably `/Int` does as well. What does Z3 integer division do?

I will go through lemmas with division later to enforce non-negativity where required.

Reply:
Z3 seems to think taht we do need positivity constraints:

This query, which is setup as `assert REQ`, then `check-set NOT (LHS == RHS)`, returns `sat`, which we need it to return `unsat`. I asked for a model, and it gave back negative numbers:

```
(declare-const a Int)
(declare-const b Int)
(declare-const c Int)
(assert (not (= b 0)))
(assert (not (= c 0)))
(assert (not (= (div (div (* a b) c) b) (div a c))))
(check-sat)
(get-model) 
```

Output:

```
sat
(
  (define-fun c () Int
    (- 2))
  (define-fun b () Int
    (- 1))
  (define-fun a () Int
    (- 3))
  (define-fun div0 ((x!0 Int) (x!1 Int)) Int
    (ite (and (= x!0 (- 1)) (= x!1 (- 1))) 1
    (ite (and (= x!0 (- 3)) (= x!1 (- 2))) 2
      (- 1))))
  (define-fun mod0 ((x!0 Int) (x!1 Int)) Int
    (ite (and (= x!0 (- 1)) (= x!1 (- 1))) 0
      1))
)
```

Reply:
Yup, well done z3, I thought /Int was floor. Will go through all /Int lemmas to assess.

Reply:
At least need positivity on `B` and `C` it looks like, so could just change side condition to say `0 <Int B` and `0 <Int C`:

This gives back `unsat`:

```
(declare-const a Int)
(declare-const b Int)
(declare-const c Int)
(assert (not (= b 0)))
(assert (not (= c 0)))
; (assert (<= 0 a))
(assert (<= 0 b))
(assert (<= 0 c))
(assert (not (= (div (div (* a b) c) b) (div a c))))
(check-sat)
; (get-model)
```

As does this:

```
(declare-const a Int)
(declare-const b Int)
(declare-const c Int)
(assert (< 0 b))
(assert (< 0 c))
(assert (not (= (div (div (* a b) c) b) (div a c))))
(check-sat)
; (get-model)
```
```